doc/wapiti.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "WAPITI" "1" "September 2024" "" ""
.
.SH "NAME"
\fBwapiti\fR \- A web application vulnerability scanner in Python
.
.SH "SYNOPSIS"
\fBwapiti\fR \-u \fIBASE_URL\fR [options]
.
.SH "DESCRIPTION"
Wapiti allows you to audit the security of your web applications\.
.
.P
It performs "black\-box" scans, i\.e\. it does not study the source code of the application but will scan the webpages of the deployed webapp, looking for scripts and forms where it can inject data\.
.
.P
Once it gets this list, Wapiti acts like a fuzzer, injecting payloads to see if a script is vulnerable\.
.
.P
Wapiti is useful only to discover vulnerabilities: it is not an exploitation tools\. Some well known applications can be used for the exploitation part like the recommended sqlmap\.
.
.SH "OPTIONS SUMMARY"
Here is a summary of options\. It is essentially what you will get when you launch Wapiti without any argument\. More detail on each option can be found in the following sections\.
.
.P
TARGET SPECIFICATION:
.
.IP "\(bu" 4
\fB\-u\fR, \fB\-\-url\fR \fIURL\fR
.
.IP "\(bu" 4
\fB\-\-data\fR \fIURL_ENCODED_DATA\fR
.
.IP "\(bu" 4
\fB\-\-scope\fR {url,page,folder,subdomain,domain,punk}
.
.IP "" 0
.
.P
ATTACK SPECIFICATION:
.
.IP "\(bu" 4
\fB\-m\fR \fIMODULES_LIST\fR
.
.IP "\(bu" 4
\fB\-\-list\-modules\fR
.
.IP "\(bu" 4
\fB\-l\fR, \fB\-\-level\fR \fILEVEL\fR
.
.IP "\(bu" 4
\fB\-\-cms\fR {drupal,joomla,prestashop,spip,wp}
.
.IP "" 0
.
.P
PROXY AND AUTHENTICATION OPTIONS:
.
.IP "\(bu" 4
\fB\-p\fR, \fB\-\-proxy\fR \fIPROXY_URL\fR
.
.IP "\(bu" 4
\fB\-\-tor\fR
.
.IP "\(bu" 4
\fB\-\-mitm\-port\fR \fIPORT\fR
.
.IP "\(bu" 4
\fB\-a\fR, \fB\-\-auth\-cred\fR \fICREDENTIALS\fR
.
.IP "\(bu" 4
\fB\-\-auth\-user\fR \fIUSERNAME\fR
.
.IP "\(bu" 4
\fB\-\-auth\-password\fR \fIPASSWORD\fR
.
.IP "\(bu" 4
\fB\-\-auth\-method\fR {basic,digest,ntlm}
.
.IP "\(bu" 4
\fB\-\-form\-cred\fR \fICREDENTIALS\fR
.
.IP "\(bu" 4
\fB\-\-form\-user\fR \fIUSERNAME\fR
.
.IP "\(bu" 4
\fB\-\-form\-password\fR \fIPASSWORD\fR
.
.IP "\(bu" 4
\fB\-\-form\-url\fR \fIURL\fR
.
.IP "\(bu" 4
\fB\-\-form\-enctype\fR \fIENCTYPE\fR
.
.IP "\(bu" 4
\fB\-\-form\-script\fR \fIFILENAME\fR
.
.IP "\(bu" 4
\fB\-\-side\-file\fR \fIFILENAME\fR
.
.IP "\(bu" 4
\fB\-c\fR, \fB\-\-cookie\fR \fICOOKIE_FILE_OR_BROWSER_NAME\fR
.
.IP "\(bu" 4
\fB\-C\fR, \fB\-\-cookie\-value\fR \fICOOKIE_VALUE\fR
.
.IP "\(bu" 4
\fB\-\-drop\-set\-cookie\fR
.
.IP "" 0
.
.P
SESSION OPTIONS:
.
.IP "\(bu" 4
\fB\-\-skip\-crawl\fR
.
.IP "\(bu" 4
\fB\-\-resume\-crawl\fR
.
.IP "\(bu" 4
\fB\-\-flush\-attacks\fR
.
.IP "\(bu" 4
\fB\-\-flush\-session\fR
.
.IP "\(bu" 4
\fB\-\-store\-session\fR \fIPATH\fR
.
.IP "\(bu" 4
\fB\-\-store\-config\fR \fIPATH\fR
.
.IP "" 0
.
.P
SCAN AND ATTACKS TUNING:
.
.IP "\(bu" 4
\fB\-s\fR, \fB\-\-start\fR \fIURL\fR
.
.IP "\(bu" 4
\fB\-x\fR, \fB\-\-exclude\fR \fIURL\fR
.
.IP "\(bu" 4
\fB\-\-swagger\fR \fIURL\fR
.
.IP "\(bu" 4
\fB\-r\fR, \fB\-\-remove\fR \fIPARAMETER\fR
.
.IP "\(bu" 4
\fB\-\-skip\fR \fIPARAMETER\fR
.
.IP "\(bu" 4
\fB\-d\fR, \fB\-\-depth\fR \fIDEPTH\fR
.
.IP "\(bu" 4
\fB\-\-max\-links\-per\-page\fR \fIMAX_LINKS_PER_PAGE\fR
.
.IP "\(bu" 4
\fB\-\-max\-files\-per\-dir\fR \fIMAX_FILES_PER_DIR\fR
.
.IP "\(bu" 4
\fB\-\-max\-scan\-time\fR \fIMAX_SCAN_TIME\fR
.
.IP "\(bu" 4
\fB\-\-max\-attack\-time\fR \fIMAX_ATTACK_TIME\fR
.
.IP "\(bu" 4
\fB\-\-max\-parameters\fR \fIMAX\fR
.
.IP "\(bu" 4
\fB\-S\fR, \fB\-\-scan\-force\fR {paranoid,sneaky,polite,normal,aggressive,insane}
.
.IP "\(bu" 4
\fB\-\-tasks\fR \fITASKS\fR
.
.IP "" 0
.
.P
ENDPOINT OPTIONS:
.
.IP "\(bu" 4
\fB\-\-external\-endpoint\fR \fIEXTERNAL_ENDPOINT_URL\fR
.
.IP "\(bu" 4
\fB\-\-internal\-endpoint\fR \fIINTERNAL_ENDPOINT_URL\fR
.
.IP "\(bu" 4
\fB\-\-endpoint\fR \fIENDPOINT_URL\fR
.
.IP "\(bu" 4
\fB\-\-dns\-endpoint\fR \fIDNS_ENDPOINT_DOMAIN\fR
.
.IP "" 0
.
.P
HTTP AND NETWORK OPTIONS:
.
.IP "\(bu" 4
\fB\-t\fR, \fB\-\-timeout\fR \fISECONDS\fR
.
.IP "\(bu" 4
\fB\-H\fR, \fB\-\-header\fR \fIHEADER\fR
.
.IP "\(bu" 4
\fB\-A\fR, \fB\-\-user\-agent\fR \fIAGENT\fR
.
.IP "\(bu" 4
\fB\-\-verify\-ssl\fR {0,1}
.
.IP "" 0
.
.P
OUTPUT OPTIONS:
.
.IP "\(bu" 4
\fB\-\-color\fR
.
.IP "\(bu" 4
\fB\-v\fR, \fB\-\-verbose\fR \fILEVEL\fR
.
.IP "\(bu" 4
\fB\-\-log\fR \fIOUTPUT_PATH\fR
.
.IP "" 0
.
.P
REPORT OPTIONS:
.
.IP "\(bu" 4
\fB\-f\fR, \fB\-\-format\fR {json,html,txt,xml}
.
.IP "\(bu" 4
\fB\-o\fR, \fB\-\-output\fR \fIOUTPUT_PATH\fR
.
.IP "\(bu" 4
\fB\-dr\fR, \fB\-\-detailed\-report\fR \fILEVEL\fR
.
.IP "" 0
.
.P
OTHER OPTIONS:
.
.IP "\(bu" 4
\fB\-\-no\-bugreport\fR
.
.IP "\(bu" 4
\fB\-\-version\fR
.
.IP "\(bu" 4
\fB\-\-update\fR [\fB\-\-wapp\-url\fR \fIWAPP_DB_URL\fR, \fB\-\-wapp\-dir\fR \fIWAPP_DB_PATH\fR]
.
.IP "\(bu" 4
\fB\-h\fR
.
.IP "" 0
.
.SH "TARGET SPECIFICATION"
.
.IP "\(bu" 4
\fB\-u\fR, \fB\-\-url\fR \fIURL\fR
.
.br
The URL that will be used as the base for the scan\. Every URL found during the scan will be checked against the base URL and the corresponding scan scope (see \-\-scope for details)\.
.
.br
This is the only required argument\. The scheme part of the URL must be either http or https\.
.
.IP "\(bu" 4
\fB\-\-data\fR \fIURL_ENCODED_DATA\fR
.
.br
If you need to attack only a specific POST request you can give this option a url\-encoded string\. It will be used as POST parameters for the URL specified by the \fB\-u\fR option\.
.
.IP "\(bu" 4
\fB\-\-scope\fR \fISCOPE\fR
.
.br
Define the scope of the scan and attacks\. Valid choices are :
.
.IP "\(bu" 4
url : will only scan and attack the exact base URL given with \-u option\.
.
.IP "\(bu" 4
page : will attack every URL matching the path of the base URL (every query string variation)\.
.
.IP "\(bu" 4
folder : will scan and attack every URL starting with the base URL value\. This base URL should have a trailing slash (no filename)\.
.
.IP "\(bu" 4
domain : will scan and attack every URL whose domain name match the one from the base URL\.
.
.IP "\(bu" 4
punk : will scan and attack every URL found whatever the domain\. Think twice before using that scope\.
.
.IP "" 0

.
.IP "" 0
.
.SH "ATTACKS SPECIFICATION"
.
.IP "\(bu" 4
\fB\-m\fR, \fB\-\-module\fR \fIMODULE_LIST\fR
.
.br
Set the list of attack modules (modules names separated with commas) to launch against the target\.
.
.br
Default behavior (when the option is not set) is to use the most common modules\.
.
.br
Common modules can also be specified using the "common" keyword\.
.
.br
If you want to use common modules along with XXE module you can pass \-m common,xxe\.
.
.br
Activating all modules can be done with the "all" keyword (not recommended though)\.
.
.br
To launch a scan without launching any attack, just give an empty value (\-m "")\.
.
.br
You can filter on http methods too (only get or post)\. For example \-m "xss:get,exec:post"\.
.
.IP "\(bu" 4
\fB\-\-list\-modules\fR
.
.br
Print the list of available Wapiti modules along with a short description then exit\.
.
.IP "\(bu" 4
\fB\-l\fR, \fB\-\-level\fR \fILEVEL\fR
.
.br
In previous versions Wapiti used to inject attack payloads in query strings even if no parameter was present in the original URL\.
.
.br
While it may be successful in finding vulnerabilities that way, it was causing too many requests for not enough success\.
.
.br
This behavior is now hidden behind this option and can be reactivated by setting \-l to 2\.
.
.br
It may be useful on CGIs when developers have to parse the query\-string themselves\.
.
.br
Default value for this option is 1\.
.
.IP "\(bu" 4
\fB\-\-cms\fR \fICMS_LIST\fR This option can only be used when the module cms is selected\. It allows to specify the CMS to scan from the list {drupal,joomla,prestashop,spip,wp}\. Multiple choices are allowed, all the CMS will be scanned if this option is not set\.
.
.IP "" 0
.
.SH "PROXY AND AUTHENTICATION"
.
.IP "\(bu" 4
\fB\-p\fR, \fB\-\-proxy\fR \fIPROXY_URL\fR
.
.br
The given URL will be used as a proxy for HTTP and HTTPS requests\. This URL can have one of the following scheme : http, https, socks\.
.
.IP "\(bu" 4
\fB\-\-tor\fR
.
.br
Make Wapiti use a Tor listener (same as \-\-proxy socks://127\.0\.0\.1:9050/)
.
.IP "\(bu" 4
\fB\-\-mitm\-port\fR \fIPORT\fR If used, this option will launch a mitmproxy instance listening on the given port instead of using an automated crawler to explore the target\. Configure your browser to use the intercepting proxy then explore the target manually\. Ctrl+C in the console when you are done\.
.
.IP "\(bu" 4
\fB\-a\fR, \fB\-\-auth\-cred\fR \fICREDENTIALS\fR
.
.br
(DEPRECATED) Set credentials to use for HTTP authentication on the target (see available methods bellow)\. Given value should be in the form login%password (% is used as a separator)
.
.IP "\(bu" 4
\fB\-\-auth\-user\fR \fIUSERNAME\fR
.
.br
Set username to use for HTTP authentication on the target (see available methods bellow)\.
.
.IP "\(bu" 4
\fB\-\-auth\-password\fR \fIPASSWORD\fR
.
.br
Set password to use for HTTP authentication on the target (see available methods bellow)\.
.
.IP "" 0

.
.IP "\(bu" 4
\fB\-\-auth\-method\fR \fITYPE\fR
.
.br
Set the authentication mechanism to use\. Valid choices are basic, digest and ntlm\. NTLM authentication may require you to install an additional Python module\.
.
.IP "\(bu" 4
\fB\-\-form\-cred\fR \fICREDENTIALS\fR
.
.br
(DEPRECATED) Set credentials to use for web form authentication on the target\. Given value should be in the form login%password (% is used as a separator)
.
.IP "\(bu" 4
\fB\-\-form\-user\fR \fIUSERNAME\fR
.
.br
Set username to use for web form authentication on the target\.
.
.IP "\(bu" 4
\fB\-\-form\-password\fR \fIPASSWORD\fR
.
.br
Set password to use for web form authentication on the target\.
.
.IP "" 0

.
.IP "\(bu" 4
\fB\-\-form\-url\fR \fIURL\fR
.
.br
If \fB\-\-form\-data\fR is not set, Wapiti will extract the login form at the given URL and fill it with the provided credentials\. Otherwise raw credentials are sent directly to the given URL\.
.
.IP "\(bu" 4
\fB\-\-form\-enctype\fR \fIENCTYPE\fR
.
.br
Send data specified with \fB\-\-form\-data\fR using the given content\-type (default is "application/x\-www\-form\-urlencoded")
.
.IP "\(bu" 4
\fB\-\-form\-script\fR \fIFILENAME\fR
.
.br
Use a custom Python authentication plugin
.
.IP "\(bu" 4
\fB\-\-side\-file\fR \fIFILENAME\fR Use a \.side file generated using Selenium IDE to perform an authenticated scan\.
.
.IP "\(bu" 4
\fB\-c\fR, \fB\-\-cookie\fR \fICOOKIE_FILE_OR_BROWSER_NAME\fR
.
.br
Load cookies from a Wapiti JSON cookie file\. See wapiti\-getcookie(1) for more information\.
.
.br
You can also import cookies from your browser by passing "chrome" or "firefox" as value (MS Edge is not supported)\.
.
.IP "\(bu" 4
\fB\-C\fR, \fB\-\-cookie\-value\fR \fICOOKIE_VALUE\fR Set cookies from a valid user cookies\. You can import all the session cookies by copying the value of the cookies sent with headers from a request sent by an authenticated user\. For example: \-\-cookie\-value "PHPSESSIONID=5f4dcc3b5aa765d61d8327deb882cf99;cookie_2=somevalue"
.
.IP "\(bu" 4
\-\-jwt` \fIJWT\fR Set JWT token from a valid user session\. You can import the JWT token value to be sent with every request for an authenticated scan\.
.
.IP "\(bu" 4
\fB\-\-drop\-set\-cookie\fR
.
.br
Ignore cookies given in HTTP responses\. Cookies that have been loaded using \fB\-c\fR will be kept\.
.
.IP "" 0
.
.SH "SESSIONS"
Since Wapiti 3\.0\.0, scanned URLs, discovered vulnerabilities and attacks status are stored in sqlite3 databases used as Wapiti session files\.
.
.br
Default behavior when a previous scan session exists for the given base URL and scope is to resume the scan and attack status\.
.
.br
Following options allows you to bypass this behavior/
.
.IP "\(bu" 4
\fB\-\-skip\-crawl\fR
.
.br
If a previous scan was performed but wasn\'t finished, don\'t resume the scan\. Attack will be made on currently known URLs without scanning more\.
.
.IP "\(bu" 4
\fB\-\-resume\-crawl\fR
.
.br
If the crawl was previously stopped and attacks started, default behavior is to skip crawling if the session is restored\.
.
.br
Use this option in order to continue the scan process while keeping vulnerabilities and attacks in the session\.
.
.IP "\(bu" 4
\fB\-\-flush\-attacks\fR
.
.br
Forget everything about discovered vulnerabilities and which URL was attacked by which module\.
.
.br
Only the scan (crawling) information will be kept\.
.
.IP "\(bu" 4
\fB\-\-flush\-session\fR
.
.br
Forget everything about the target for the given scope\.
.
.IP "\(bu" 4
\fB\-\-store\-session\fR \fIPATH\fR
.
.br
Specify an alternative path for storing session (\.db and \.pkl) files\.
.
.IP "\(bu" 4
\fB\-\-store\-config\fR \fIPATH\fR
.
.br
Specify an alternative path for storing particular module (\fBapps\.json\fR and \fBnikto_db\fR) files\.
.
.IP "" 0
.
.SH "SCAN AND ATTACKS TUNING"
.
.IP "\(bu" 4
\fB\-s\fR, \fB\-\-start\fR \fIURL\fR
.
.br
If for some reasons, Wapiti doesn\'t find any (or enough) URLs from the base URL you can still add URLs to start the scan with\.
.
.br
Those URLs will be given a depth of 0, just like the base URL\.
.
.br
This option can be called several times\.
.
.br
You can also give it a filename and Wapiti will read URLs from the given file (must be UTF\-8 encoded), one URL per line\.
.
.IP "\(bu" 4
\fB\-x\fR, \fB\-\-exclude\fR \fIURL\fR
.
.br
Prevent the given URL from being scanned\. Common use is to exclude the logout URL to prevent the destruction of session cookies (if you specified a cookie file with \-\-cookie)\.
.
.br
This option can be applied several times\. Excluded URL given as a parameter can contain wildcards for basic pattern matching\.
.
.IP "\(bu" 4
\fB\-\-swagger\fR \fIURL\fR
.
.br
Extract API requests from the specified Swagger file\. Extracted requests are added to the crawler\.
.
.IP "\(bu" 4
\fB\-r\fR, \fB\-\-remove\fR \fIPARAMETER\fR
.
.br

wapitiCore/main/wapiti.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# This file is part of the Wapiti project (https://wapiti-scanner.github.io)
# Copyright (C) 2006-2023 Nicolas SURRIBAS
# Copyright (C) 2020-2024 Cyberwatch
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
import asyncio
import codecs
import os
import signal
import sys
from importlib import import_module
from inspect import getdoc
from sqlite3 import OperationalError
from urllib.parse import urlparse

import httpx
from httpx import RequestError
from wapitiCore.attack.attack import (all_modules, common_modules)
from wapitiCore.controller.wapiti import InvalidOptionValue, module_to_class_name, Wapiti
from wapitiCore.main.banners import print_banner
from wapitiCore.parsers.commandline import parse_args
from wapitiCore.main.log import logging
from wapitiCore.net.classes import HttpCredential, FormCredential, RawCredential
from wapitiCore.net.auth import (async_try_form_login, load_form_script, check_http_auth, login_with_raw_data,
                                 authenticate_with_side_file)
from wapitiCore.net import Request
from wapitiCore.report import GENERATORS
from wapitiCore.parsers.swagger import Swagger

global_stop_event = asyncio.Event()


def inner_ctrl_c_signal_handler():
    logging.info("Waiting for running crawler tasks to finish, please wait.")
    global_stop_event.set()


def stop_attack_process():
    logging.info("Waiting for all payload tasks to finish for current resource, please wait.")
    global_stop_event.set()


def fix_url_path(url: str):
    """Fix the url path if it's not defined"""
    return url if urlparse(url).path else url + '/'


def is_valid_url(url: str):
    """Verify if the url provided has the right format"""
    try:
        parts = urlparse(url)
    except ValueError:
        logging.error('ValueError')
        return False
    else:
        if parts.scheme in ("http", "https") and parts.netloc:
            return True
    logging.error(f"Error: {url} is not a valid URL")
    return False


def is_valid_endpoint(url_type, url: str):
    """Verify if the url provided has the right format"""
    try:
        parts = urlparse(url)
    except ValueError:
        logging.error('ValueError')
        return False
    else:
        if parts.params or parts.query or parts.fragment:
            logging.error(f"Error: {url_type} must not contain params, query or fragment!")
            return False
        if parts.scheme in ("http", "https") and parts.netloc:
            return True
    logging.error(f"Error: {url_type} must contain scheme and host")
    return False


def is_mod_cms_set(args):
    if args.modules and "cms" in args.modules:
        return True
    logging.error("Error: Invalid option --cms, module cms is required when this option is used")
    return False


def is_mod_wapp_or_update_set(args):
    if (args.modules and "wapp" in args.modules) or args.update:
        return True
    return False


def validate_cms_choices(cms_value):
    allowed_cms = ["drupal", "joomla", "prestashop", "spip", "wp"]
    cms_list = cms_value.split(',')
    for cms in cms_list:
        if cms not in allowed_cms:
            logging.error(f"Invalid CMS choice: {cms_value}. Choose from {', '.join(allowed_cms)}")
            return False
    return True


def ping(url: str):
    try:
        httpx.get(url, timeout=5)
    except RequestError:
        return False
    return True


async def wapiti_main():
    print_banner()
    args = parse_args()

    if args.tasks < 1:
        logging.error("Number of concurrent tasks must be 1 or above!")
        sys.exit(2)

    if args.scope == "punk":
        print("[*] Do you feel lucky punk?")

    if args.list_modules:
        print("[*] Available modules:")
        for module_name in sorted(all_modules):
            try:
                mod = import_module("wapitiCore.attack.mod_" + module_name)
                class_name = module_to_class_name(module_name)
                is_common = " (used by default)" if module_name in common_modules else ""
                print(f"\t{module_name}{is_common}")
                print("\t\t" + getdoc(getattr(mod, class_name)))
                print('')
            except ImportError:
                continue
        sys.exit()

    url = fix_url_path(args.base_url)
    if args.data:
        base_request = Request(
            url,
            method="POST",
            post_params=args.data
        )
    else:
        base_request = Request(url)

    parts = urlparse(url)
    if not parts.scheme or not parts.netloc:
        logging.error("Invalid base URL was specified, please give a complete URL with protocol scheme.")
        sys.exit()

    wap = Wapiti(base_request, scope=args.scope, session_dir=args.store_session, config_dir=args.store_config)

    if args.log:
        wap.set_logfile(args.log)

    if "proxy" in args:
        wap.set_proxy(args.proxy)

    if args.tor:
        wap.set_proxy("socks5://127.0.0.1:9050/")

    if args.update:
        await wap.init_persister()
        logging.log("GREEN", "[*] Updating modules")
        if args.wapp_url:
            attack_options = {"level": args.level, "timeout": args.timeout, "wapp_url": fix_url_path(args.wapp_url)}
        elif args.wapp_dir:
            attack_options = {"level": args.level, "timeout": args.timeout, "wapp_dir": args.wapp_dir}
        else:
            attack_options = {"level": args.level, "timeout": args.timeout,\
                              "wapp_url": "https://raw.githubusercontent.com/wapiti-scanner/wappalyzer/main/"}
        wap.set_attack_options(attack_options)
        try:
            await wap.update(args.modules)
            sys.exit()
        except InvalidOptionValue as invalid_option:
            logging.error(invalid_option)
            raise
        except ValueError as e:
            logging.error(f"Value error: {e}")
            raise

    if args.swagger_uri:
        swagger = Swagger(swagger_url=args.swagger_uri, base_url=url)
        nb_out = 0
        for request in swagger.get_requests():
            if wap.target_scope.check(request):
                wap.add_start_url(request)
            else:
                nb_out += 1
        if nb_out > 0:
            logging.warning(f"[!] {nb_out} out of scope requests from the Swagger file are not added.")

    try:
        for start_url in args.starting_urls:
            if start_url.startswith(("http://", "https://")):
                wap.add_start_url(Request(start_url))
            elif os.path.isfile(start_url):
                try:
                    with codecs.open(start_url, encoding="UTF-8") as urlfd:
                        for urlline in urlfd:
                            urlline = urlline.strip()
                            if urlline.startswith(("http://", "https://")):
                                wap.add_start_url(Request(urlline))
                except UnicodeDecodeError as exception:
                    logging.error("Error: File given with the -s option must be UTF-8 encoded !")
                    raise InvalidOptionValue("-s", start_url) from exception
            else:
                raise InvalidOptionValue('-s', start_url)

        for exclude_url in args.excluded_urls:
            if exclude_url.startswith(("http://", "https://")):
                wap.add_excluded_url(exclude_url)
            else:
                raise InvalidOptionValue("-x", exclude_url)

        if "mitm_port" in args:
            wap.set_intercepting_proxy_port(args.mitm_port)

        wap.set_headless(args.headless)
        wap.set_wait_time(args.wait_time)

        if "side_file" in args:
            if os.path.isfile(args.side_file):
                wap.crawler_configuration.cookies = await authenticate_with_side_file(
                    wap.crawler_configuration, args.side_file, args.headless
                )

        if "cookie" in args:
            if os.path.isfile(args.cookie):
                wap.set_cookie_file(args.cookie)
            elif args.cookie.lower() in ("chrome", "firefox"):
                wap.load_browser_cookies(args.cookie)
            else:
                raise InvalidOptionValue("-c", args.cookie)

        if args.drop_set_cookie:
            wap.set_drop_cookies()

        if "http_credentials" in args:
            # This option is deprecated, but we still support it
            # Should be removed in the future
            if "%" in args.http_credentials:
                username, password = args.http_credentials.split("%", 1)
                wap.set_http_credentials(HttpCredential(username, password, args.auth_method))
            else:
                raise InvalidOptionValue("-a", args.http_credentials)
        elif "http_user" in args and "http_password" in args:
            wap.set_http_credentials(HttpCredential(args.http_user, args.http_password, args.auth_method))

        if ("http_user" in args and "http_password" not in args) or \
           ("http_user" not in args and "http_password" in args):
            raise InvalidOptionValue("--auth-user and --auth-password",
                                     "Both options are required when one is used")

        for bad_param in args.excluded_parameters:
            wap.add_bad_param(bad_param)

        wap.set_max_depth(args.depth)
        wap.set_max_files_per_dir(args.max_files_per_dir)
        wap.set_max_links_per_page(args.max_links_per_page)
        wap.set_scan_force(args.scan_force)
        wap.set_max_scan_time(args.max_scan_time)
        wap.set_max_attack_time(args.max_attack_time)

        # should be a setter
        wap.verbosity(args.verbosity)
        if args.detailed_report_level:
            wap.set_detail_report(args.detailed_report_level)
        if args.color:
            wap.set_color()
        wap.set_timeout(args.timeout)
        wap.set_modules(args.modules)

        if args.no_bugreport:
            wap.set_bug_reporting(False)

        if "user_agent" in args:
            wap.add_custom_header("User-Agent", args.user_agent)

        if "cookie_value" in args:
            wap.add_custom_header("Cookie", args.cookie_value)

        if "jwt" in args:
            wap.add_custom_header("Authorization", f"Bearer {args.jwt}")

        for custom_header in args.headers:
            if ":" in custom_header:
                hdr_name, hdr_value = custom_header.split(":", 1)
                wap.add_custom_header(hdr_name.strip(), hdr_value.strip())

        if "output" in args:
            wap.set_output_file(args.output)

        if args.format not in GENERATORS:
            raise InvalidOptionValue("-f", args.format)

        wap.set_report_generator_type(args.format)

        wap.set_verify_ssl(bool(args.check_ssl))

        attack_options = {
            "level": args.level,
            "timeout": args.timeout,
            "tasks": args.tasks,
            "headless": wap.headless_mode,
            "excluded_urls": wap.excluded_urls,
            "max_attack_time": args.max_attack_time,
        }

        if "dns_endpoint" in args:
            attack_options["dns_endpoint"] = args.dns_endpoint

        if "endpoint" in args:
            endpoint = fix_url_path(args.endpoint)
            if is_valid_endpoint('ENDPOINT', endpoint):
                attack_options["external_endpoint"] = endpoint
                attack_options["internal_endpoint"] = endpoint
            else:
                raise InvalidOptionValue("--endpoint", args.endpoint)

        if "external_endpoint" in args:
            external_endpoint = fix_url_path(args.external_endpoint)
            if is_valid_endpoint('EXTERNAL ENDPOINT', external_endpoint):
                attack_options["external_endpoint"] = external_endpoint
            else:
                raise InvalidOptionValue("--external-endpoint", external_endpoint)

        if "internal_endpoint" in args:
            internal_endpoint = fix_url_path(args.internal_endpoint)
            if is_valid_endpoint('INTERNAL ENDPOINT', internal_endpoint):
                if ping(internal_endpoint):
                    attack_options["internal_endpoint"] = internal_endpoint
                else:
                    logging.error("Error: Internal endpoint URL must be accessible from Wapiti!")
                    raise InvalidOptionValue("--internal-endpoint", internal_endpoint)
            else:
                raise InvalidOptionValue("--internal-endpoint", internal_endpoint)

        if args.cms:
            allowed_cms = ["drupal", "joomla", "prestashop", "spip", "wp"]
            if not is_mod_cms_set(args):
                raise InvalidOptionValue("--cms", "module cms is required when --cms is used")
            if not validate_cms_choices(args.cms):
                raise InvalidOptionValue(
                    "--cms", f"Invalid CMS choice: {args.cms}. Choose from {', '.join(allowed_cms)}"
                )
            attack_options["cms"] = args.cms

        if args.modules and "cms" in args.modules and not args.cms:
            attack_options["cms"] = "drupal,joomla,prestashop,spip,wp"

        if args.wapp_url:
            if not is_mod_wapp_or_update_set(args):
                raise InvalidOptionValue("--wapp-url", "module wapp or --update option is required when --wapp-url is "
                                                       "used")
            url_value = fix_url_path(args.wapp_url)
            if is_valid_url(url_value):
                attack_options["wapp_url"] = url_value
            else:
                raise InvalidOptionValue(
                    "--wapp-url", url_value
                )

        if args.wapp_dir:
            if not is_mod_wapp_or_update_set(args):
                raise InvalidOptionValue("--wapp-url", "module wapp or --update option is required when --wapp-url is "
                                                       "used")
            dir_value = args.wapp_dir
            if os.path.isdir(dir_value):
                attack_options["wapp_dir"] = dir_value
            else:
                raise InvalidOptionValue(
                    "--wapp-dir", dir_value
                )

        if args.skipped_parameters:
            attack_options["skipped_parameters"] = set(args.skipped_parameters)

        wap.set_attack_options(attack_options)

        await wap.init_persister()
        if args.flush_attacks:
            await wap.flush_attacks()

        if args.flush_session:
            await wap.flush_session()

    except InvalidOptionValue as msg:
        logging.error(msg)
        sys.exit(2)

    assert os.path.exists(wap.history_file)

    if "http_credentials" in args or ("http_user" in args and "http_password" in args):
        if not await check_http_auth(wap.crawler_configuration):
            logging.warning("[!] HTTP authentication failed, a 4xx status code was received")
            return

    form_credential = None
    if "form_credentials" in args:
        if "form_url" not in args:
            raise InvalidOptionValue(
                "--form-url",
                "This option is required when --form-user and --form-password or form-cred is used")
        # This option is deprecated, but we still support it
        # Should be removed in the future
        username, password = None, None
        if "%" in args.form_credentials:
            username, password = args.form_credentials.split("%", 1)
            form_credential = FormCredential(
            username,
            password,
            args.form_url,
        )
        else:
            raise InvalidOptionValue("--form-cred", args.form_credentials)
    elif "form_user" in args and "form_password" in args:
        if "form_url" not in args:
            raise InvalidOptionValue(
                "--form-url",
                "This option is required when --form-user and --form-password or form-cred is used")
        form_credential = FormCredential(
            args.form_user,
            args.form_password,
            args.form_url,
        )

    if ("form_user" in args and "form_password" not in args) or ("form_user" not in args and "form_password" in args):
        raise InvalidOptionValue("--form-user and --form-password", "Both options are required when one is used")

    if "form_script" in args:
        await load_form_script(
            args.form_script,
            wap.crawler_configuration,
            form_credential,  # Either None or filled
            wap.headless_mode
        )
    elif "form_data" in args:
        if "form_url" not in args:
            raise InvalidOptionValue("--form-url", "This option is required when --form-data is used")

        raw_credential = RawCredential(
            args.form_data,
            args.form_url,
            args.form_enctype
        )
        await login_with_raw_data(wap.crawler_configuration, raw_credential)
    elif form_credential:
        is_logged_in, form, excluded_urls = await async_try_form_login(
            wap.crawler_configuration,
            form_credential,
            wap.headless_mode,
        )
        wap.set_auth_state(is_logged_in, form, form_credential.url)
        for url in excluded_urls:
            wap.add_excluded_url(url)

    loop = asyncio.get_event_loop()

    try:
        if not args.skip_crawl:
            if await wap.have_attacks_started() and not args.resume_crawl:
                pass
            else:
                if await wap.has_scan_started():
                    logging.info("[*] Resuming scan from previous session, please wait")

                await wap.load_scan_state()
                loop.add_signal_handler(signal.SIGINT, inner_ctrl_c_signal_handler)
                await wap.browse(global_stop_event, parallelism=args.tasks)
                loop.remove_signal_handler(signal.SIGINT)
                await wap.save_scan_state()

        if args.max_parameters:
            count = await wap.persister.remove_big_requests(args.max_parameters)
            logging.info(
                f"[*] {count} URLs and forms having more than {args.max_parameters} parameters were removed."
            )

        logging.info(f"[*] Wapiti found {await wap.count_resources()} URLs and forms during the scan")
        loop.add_signal_handler(signal.SIGINT, stop_attack_process)
        await wap.attack(global_stop_event)
        loop.remove_signal_handler(signal.SIGINT)

    except OperationalError:
        logging.error(
            "[!] Can't store information in persister. SQLite database must have been locked by another process"
        )
        logging.error("[!] You should unlock and launch Wapiti again.")
    except SystemExit:
        pass


def wapiti_asyncio_wrapper():
    asyncio.run(wapiti_main())
